projects/beta_49_ssl_handshake_version_spike.kp/knowledge.md

```diff
@@ -0,0 +1,246 @@
+
+
+```python
+from matplotlib import pyplot as plt
+import seaborn as sns
+
+%matplotlib inline
+
+plt.title("Number of SSL_HANDSHAKE_VERSION Measurements per Ping")
+plt.xlabel("SSL_HANDSHAKE_VERSION Measurements")
+plt.ylabel("Ping Count")
+
+plt.hist(counts, bins=50)
+plt.legend()
+```
+
+
+![png](images/output_9_1.png)
+
+
+## We are looking at duplicate pings
+
+The plot below shows number of pings and number of distinct DocIDs for this single client per day (looking around the problem time period). We are seeing about 10 repeated docIds, all with high counts of SSL_HANDSHAKE_VERSION.
+
+We have started deduping server-side so some of these pings would be removed; deduping will improve over the next few months so that we recieve less than 10% of dupes.
+
+
+```python
+# What does this client's distribution look like across a the bad days?
+# We know bad days from the aggregates data, see
+# https://github.com/fbertsch/telemetry-repro/blob/2bf7e062719ae498c12f15456e94af9ee7758093/output.txt
+
+begin = '20170305'
+end = '20170320'
+
+cid_pings = Dataset \
+    .from_source('telemetry') \
+    .where (
+        appUpdateChannel = 'beta',
+        appVersion = lambda x: x.startswith('49'),
+        submissionDate = lambda x: begin <= x <= end
+    ).records(sc) \
+    .filter(lambda x: x.get('clientId') == cid)
+
+props = {
+    'submission_date': 'meta/submissionDate',
+    'version': 'version',
+    'build_id': 'buildId',
+    'document_id': 'meta/documentId',
+    'timestamp': 'meta/Timestamp',
+    'SSL_HANDSHAKE_VERSION': 'payload/histograms/SSL_HANDSHAKE_VERSION'
+}
+    
+data = get_pings_properties(cid_pings, props)
+
+"Total of {} pings from client".format(data.count())
+```
+    fetching 223752.06889MB in 118090 files...
+
+
+
+
+
+
+    'Total of 1282154 pings from client'
+
+
+
+
+```python
+counts = data.map(
+    lambda x: (datetime.strptime(x["submission_date"], '%Y%m%d'), 1)
+).countByKey()
+
+pings_counts = [(d,v) for d,v in counts.iteritems() if isinstance(d, datetime)]
+
+doc_counts = data.map(
+    lambda x: (datetime.strptime(x["submission_date"], '%Y%m%d'), {x['document_id']})
+).reduceByKey(
+    lambda x, y: x | y
+).map(
+    lambda x: (x[0], len(x[1]))
+).collect()
+```
+
+```python
+#How many pings and how many docs
+ping_df = pd.DataFrame(pings_counts, columns=["date", "ping_count"]).set_index(["date"])
+doc_df = pd.DataFrame(doc_counts, columns=["date", "doc_count"]).set_index(["date"])
+joined_df = ping_df.join(doc_df)
+
+joined_df.plot(figsize=(17, 7))
+plt.xticks(np.arange(min(joined_df.index), max(joined_df.index) + timedelta(3), 3, dtype="datetime64[D]"))
+
+plt.title("Pings Count and Distint DocID Count per Day")
+plt.ylabel("Count")
+plt.xlabel("Date")
+
+ax = plt.subplot(111)
+ax.set_yscale("log", nonposy='clip')
+
+plt.grid(True)
+plt.show()
+```
+
+
+![png](images/output_13_0.png)
```

Comment:
Need analysis/interpretation/conclusions under here.

Is it ongoing? How was it affected by the beginning of "discarding duplicates" action at ingestion on May 10? Do we have mechanisms for detecting/mitigating this besides waiting for people to notice anomalies (First mconley with Aurora 51, now ekr with Beta 49)? In the broader scheme, does this "matter"?

Wild speculation of what could be the cause is also a lot of fun to read :)